Improved doc of FCAlgorithm

mbaudin47 opened the pull request:
The goal of this PR is to provide a consistent documentation of the following API help pages:
- `FunctionalChaosAlgorithm` (see [FunctionalChaosAlgorithm](http://openturns.github.io/openturns/latest/user_manual/response_surface/_generated/openturns.FunctionalChaosAlgorithm.html) for the current API help page)
- `FunctionalChaosRandomVector` (see [FunctionalChaosRandomVector](http://openturns.github.io/openturns/latest/user_manual/response_surface/_generated/openturns.FunctionalChaosRandomVector.html) for the current API help page)

The following changes were made:
- `FunctionalChaosAlgorithm`: Provides the default settings used for unspecified input arguments.
- `FunctionalChaosRandomVector`: Provides an example to the  class (the current doc  does not provide any example).

Moreover, the PR provides a consistent review of the mathematical equations of the theory help pages of:
- functional_chaos
- orthogonal_polynomials
- chaos_basis

In these pages, I added several pointers to theory and API help pages. 

This is a joint work with @adutfoy that suggested changes integrated in https://github.com/mbaudin47/openturns/pull/14.

josephmure commented on python/doc/theory/meta_modeling/chaos_basis.rst:
```diff
@@ -3,132 +3,191 @@
+  \end{cases}
+
+Under some hypotheses on the function :math:`h` and the functional basis
+:math:`\{\psi_k\}_{k \geq 0}`, it can be proved that :math:`\vect{Y}` can
```
```suggestion
:math:`\{\psi_k\}_{k \geq 0}`, it can be proved that :math:`Y` can
```
Y is scalar.

josephmure commented on python/doc/theory/meta_modeling/chaos_basis.rst:
```diff
@@ -3,132 +3,191 @@
 
-      Y \, \,  \equiv \, \,  h(\underline{X}) \, \, = \, \, \sum_{j=0}^{\infty} \; a_{j} \; \psi_{j}(\underline{X})
+    Y = h(\vect{Z})
+    =  \sum_{k \geq 0} \; a_k \; \psi_k(\vect{Z}).
```
```suggestion
    =  \sum_{k \geq 0} \; a_k \; \psi_k(\vect{Z}),
```
This is not the end of the sentence.

josephmure commented on python/doc/theory/meta_modeling/chaos_basis.rst:
```diff
@@ -3,132 +3,191 @@
 
-  .. math:: \langle \psi_{j}(\underline{X}) \; , \; \psi_{k}(\underline{X}) \rangle \, \, \equiv \, \, \Expect{\psi_{j}(\underline{X}) \; \psi_{k}(\underline{X})} \, \, = \, \, \delta_{j,k}
+The goal of the next two sections is to present methods to create
+the basis :math:`\{\psi_k \}_{k \geq 0}` in two different cases:
```
```suggestion
the basis :math:`\{\psi_k \}_{k \geq 0}` in two different cases:

```
To make the bullet points work properly.

josephmure commented on python/doc/theory/meta_modeling/chaos_basis.rst:
```diff
@@ -3,132 +3,191 @@
+The goal of the next two sections is to present methods to create
+the basis :math:`\{\psi_k \}_{k \geq 0}` in two different cases:
+- if the input random vector :math:`\vect{X}` has independent marginals,
+- if the input random vector :math:`\vect{X}` has a non trivial copula.
```
```suggestion
- if the input random vector :math:`\vect{X}` has a non-trivial copula.
```

josephmure commented on python/doc/theory/meta_modeling/chaos_basis.rst:
```diff
@@ -3,132 +3,191 @@
+:math:`T_i`, usually referred to as an *isoprobabilistic
+transformation* (see :ref:`isoprobabilistic_transformation` for
+details on this topic).
+The set of standardized random variables reads:
```
```suggestion
The set of standardized random variables is:
```

josephmure commented on python/doc/theory/meta_modeling/chaos_basis.rst:
```diff
@@ -3,132 +3,191 @@
+let us consider a family :math:`\left\{\pi^{(i)}_j, j \in \Nset\right\}` of
+orthonormal polynomials with respect to :math:`f_{Z_i}` where
+:math:`j` represents the degree of the polynomial :math:`\pi^{(i)}_j`.
+By hypothesis, the polynomials are orthonormal, which implies:
```
```suggestion
By hypothesis, the polynomials are orthonormal:
```
The equation that follows is not implied, it is the definition of orthonormality.

josephmure commented on python/doc/theory/meta_modeling/chaos_basis.rst:
```diff
@@ -3,132 +3,191 @@
-  set of orthonormal multivariate polynomials
-  :math:`\{\psi_{\idx}, \idx \in \Nset^n\}` defined by:
+for any :math:`j, k \in \Nset`.
+More details orthogonal polynomials are presented in
```
```suggestion
More details about orthogonal polynomials are presented in
```

josephmure commented on python/doc/theory/meta_modeling/chaos_basis.rst:
```diff
@@ -3,132 +3,191 @@
 
-| In case of *dependent* variables, it is possible to build up an
-  orthonormal basis as follows:
+In case of *dependent* variables, it is possible to build up an
```
```suggestion
In the case of *dependent* variables, it is possible to build up an
```

josephmure commented on python/doc/theory/meta_modeling/chaos_basis.rst:
```diff
@@ -3,132 +3,191 @@
+dealing with independent random variables, each basis element may be recast as in
+:eq:`3.010six`.
+Computing :math:`\psi_{\idx}(\vect{z})` can be expensive.
+This is the reason why we present below other methods for specific types
```
below? What are you referring to?

mbaudin47 replied:
I mean "the next paragraph". So I replace "below" with "now".

mbaudin47 replied:
Fixed at ea9fdfcdb.

josephmure commented on python/doc/theory/meta_modeling/chaos_basis.rst:
```diff
@@ -3,132 +3,191 @@
+independent components using a suitable mapping
+:math:`T : \vect{X} \mapsto \vect{Z}` such as the Nataf transformation
+for example.
+The so-called Rosenblatt transformation may also be applied in case
+of a Gaussian copula.
```
```suggestion
The so-called Rosenblatt transformation may also be applied in the case
of a Normal copula.
```

josephmure commented on python/doc/theory/meta_modeling/chaos_basis.rst:
```diff
@@ -3,132 +3,191 @@
 In a deterministic setting (i.e. when the input parameters are
 considered to be deterministic), it is of common practice to substitute
 the model function :math:`h` by a polynomial approximation over its
 whole domain of definition as shown in . Actually this approach is
```
as shown where?

mbaudin47 replied:
I do not know. I remove this part of the sentence.

josephmure commented on python/doc/theory/meta_modeling/chaos_basis.rst:
```diff
@@ -3,132 +3,191 @@
 strictly equivalent to:
 
-#. Regarding the input parameters as random uniform random variables
+- Regarding the input parameters as random uniform random variables
 
-#. Expanding any quantity of interest provided by the model onto a PC
-   expansion made of Legendre polynomials
+- Expanding any quantity of interest provided by the model onto a PC
+  expansion made of Legendre polynomials
```
I wonder if a single sentence would not be clearer in this case.

mbaudin47 replied:
This is a method defined by two steps. This is why I think that the wording is correct. By the way, removing the enumerated list was not a good idea: I fix this.

josephmure replied:
ok

josephmure reviewed (changes requested):
Thanks for cleaning this section of the doc!

josephmure commented on python/src/FunctionalChaosAlgorithm_doc.i.in:
```diff
@@ -18,158 +18,238 @@ Available constructors:
 
-**Step 2 - Truncation strategy of the multivariate orthonormal basis**: a
+**Step 2 - Truncation of the multivariate orthonormal basis**: a
 strategy must be chosen for the selection of the different terms of the
```
Maybe mention it is the adaptive strategy. I can't make the suggestion, Github won't let me.

mbaudin47 replied:
The adaptive strategy is usually considered as a method to compute the coefficients. This is why the current text seems to be correct to me.

josephmure replied:
OK

josephmure commented on python/src/FunctionalChaosAlgorithm_doc.i.in:
```diff
@@ -18,158 +18,238 @@ Available constructors:
-strategy must be chosen for the estimation of te coefficients :math:`\alpha_k`. 
-The vector :math:`\vect{\alpha} = (\alpha_k)_{k \in K}` is equivalently defined by:
+**Step 3 -  Evaluation of the coefficients**: a
+strategy must be chosen for the estimation of the coefficients :math:`a_k`. 
```
```suggestion
*projectionStrategy* must be chosen for the estimation of the coefficients :math:`a_k`. 
```

josephmure commented on python/src/FunctionalChaosAlgorithm_doc.i.in:
```diff
@@ -181,14 +261,14 @@ Define the evaluation strategy of the  coefficients:
 
->>> algo = ot.FunctionalChaosAlgorithm(X, Y, distribution, adaptiveStrategy,
-...                                    projectionStrategy)
+>>> algo = ot.FunctionalChaosAlgorithm(inputSample, outputSample, distribution, \
```
Maybe rename "inputSample" as "x" and "outputSample" as "y" to save space and allow this to fit on a single line?

mbaudin47 replied:
These are the names used in the description of the arguments of the constructor of the class. These names cannot be changed because they must match the names used in the C++. This is because SWIG uses these names when generating the part of the API doc which is automatically generated. Hence, I think that this example is better as is.

josephmure replied:
Then I think you should at least align the arguments vertically.

mbaudin47 replied:
That is true. I extracted the code and formatted it using black. Fixed at edf67cfd0

josephmure commented on python/src/FunctionalChaosAlgorithm_doc.i.in:
```diff
@@ -202,7 +282,58 @@ Test it:
+settings are used when the information is not provided by the user.
+The simplest is to set only the input and output samples.
+In this case, the distribution and its parameters are estimated from
+the `inputSample` using the :class:`~openturns.MetaModelAlgorithm.BuildDistribution` class.
```
```suggestion
the *inputSample* using the :class:`~openturns.MetaModelAlgorithm.BuildDistribution` class.
```
inputSample is a parameter so we use italics (at least this seems to be the convention in other API doc pages).

josephmure commented on python/src/FunctionalChaosAlgorithm_doc.i.in:
```diff
@@ -202,7 +282,58 @@ Test it:
+
+>>> algo = ot.FunctionalChaosAlgorithm(inputSample, outputSample)
+
+In many cases, the distribution is known and the best is to use this information
```
```suggestion
In many cases, the distribution is known and it is best to use this information
```

josephmure commented on python/src/FunctionalChaosAlgorithm_doc.i.in:
```diff
@@ -202,7 +282,58 @@ Test it:
+>>> polynomialsList = []
+>>> for i in range(inputDimension):
+...     marginalDistribution = distribution.getMarginal(i)
+...     marginalPolynomial = ot.StandardDistributionPolynomialFactory( \
+...         marginalDistribution
+...     )
```
This should fit on a single line.

mbaudin47 replied:
If we do it, the line has 92 characters, 3 characters more than the expected 89 maximum number of characters. This is why I leave it as is.

josephmure replied:
Couldn't we reduce the number of characters of the "marginalPolynomial" variable then? Use somthing like "marginalPolynom" This should remove 3 characters!

mbaudin47 replied:
```suggestion
...     marginalPolynomial = ot.StandardDistributionPolynomialFactory(marginalDistribution)
```

josephmure commented on python/src/FunctionalChaosAlgorithm_doc.i.in:
```diff
@@ -202,7 +282,58 @@ Test it:
+...     ot.LARS(), \
+...     ot.CorrectedLeaveOneOut()
+... )
+>>> projectionStrategy = ot.LeastSquaresStrategy(inputSample, outputSample, \
+...                                              selection)
+>>> algo = ot.FunctionalChaosAlgorithm(inputSample, outputSample, distribution, \
+...                                    adaptiveStrategy, projectionStrategy)"
```
Why so many lines?

mbaudin47 replied:
You are right for the line 333: I fix this. For the lines 335-336, the last character will be beyond the 89th character, which is not easy to read and maintain.

mbaudin47 replied:
Partially fixed at ea9fdfcdb

josephmure replied:
If you can't fit everything on a single line, I would then align the arguments vertically.

josephmure replied:
I am OK with the suggestion you just made, so you can resolve this.

josephmure commented on python/src/FunctionalChaosAlgorithm_doc.i.in:
```diff
@@ -202,7 +282,58 @@ Test it:
+...     polynomialsList.append(marginalPolynomial)
+>>> basis = ot.OrthogonalProductPolynomialFactory(polynomialsList, enumerateFunction)
+>>> adaptiveStrategy = ot.FixedStrategy(basis, totalSize)
+>>> algo = ot.FunctionalChaosAlgorithm(inputSample, outputSample, distribution, \
+...                                    adaptiveStrategy)
```
Also this.

mbaudin47 replied:
A one-liner requires 98 characters.

josephmure replied:
I am not sure about your requirements that the variable names in the example should match the parameter names in the docstring.

mbaudin47 replied:
```suggestion
>>> algo = ot.FunctionalChaosAlgorithm(
...     inputSample, outputSample, distribution, adaptiveStrategy
.... )
```

josephmure commented on python/src/FunctionalChaosAlgorithm_doc.i.in:
```diff
@@ -280,6 +411,6 @@ residual : float
 Parameters
 ----------
 strategy : :class:`~openturns.ProjectionStrategy`
```
```suggestion
projectionStrategy : :class:`~openturns.ProjectionStrategy`
```
For the sake of consistency:
![image](https://github.com/openturns/openturns/assets/31989332/ed205376-6f2b-4679-bee2-86b04275c974)

josephmure commented on python/src/FunctionalChaosRandomVector_doc.i.in:
```diff
@@ -9,10 +9,102 @@ Parameters
+For example, we can get the output mean or the output covariance matrix
+using the coefficients of the expansion.
+
+Moreover, we can use this class to simulation random observations
```
```suggestion
Moreover, we can use this class to simulate random observations
```

josephmure commented on python/src/FunctionalChaosRandomVector_doc.i.in:
```diff
@@ -9,10 +9,102 @@ Parameters
+
+This equation can be used to simulate independent random observations
+from the PCE.
+This can be done by first simulating independent observations from
+the distribution of the standardized random vector $\vect{Z}$,
+then push forward these observations through the expansion.
```
```suggestion
This can be done by simulating independent observations from
the distribution of the standardized random vector :math:`\vect{Z}`,
which are then pushed forward through the expansion.
```

josephmure commented on python/src/FunctionalChaosRandomVector_doc.i.in:
```diff
@@ -9,10 +9,102 @@ Parameters
+>>> functionalChaosResult = algo.getResult()
+
+Secondly, we get the probabilistic properties of the PCE.
+We can get the output mean :meth:`~openturns.FunctionalChaosRandomVector.getMean`, which produces an estimate of the
```
```suggestion
We can get an estimate of the
```

josephmure commented on python/src/FunctionalChaosRandomVector_doc.i.in:
```diff
@@ -9,10 +9,102 @@ Parameters
+>>> print(mean)
+[0.301168]
+
+We can get the output covariance matrix, which is an estimate of the
+mean of the output of the physical model using :meth:`~openturns.FunctionalChaosRandomVector.getCovariance`.
```
```suggestion
We can get an estimate of the covariance matrix
of the output of the physical model.
```

josephmure commented on python/src/FunctionalChaosRandomVector_doc.i.in:
```diff
@@ -9,10 +9,102 @@ Parameters
+>>> print(covariance)
+[[ 0.0663228 ]]
+
+The `getSample()` generates observations from the PCE random
```
```suggestion
We can finally generate observations from the PCE random
```

mbaudin47 commented on python/doc/theory/meta_modeling/chaos_basis.rst:
```diff
@@ -3,132 +3,191 @@
+   :math:`\Var{h(\vect{Z})} < \infty`;
 
-  .. math::
+-  :math:`\vect{Z}` has independent components.
```
Not all transformations $T$ guarantee that $\boldsymbol{Z}$ has independent marginals.

josephmure replied:
OK, but I thought it was implied that we would use such a transformation.

adutfoy commented:
Thanks for this work! The documentation is really better than the previous one.
I would like to commit in [mbaudin47:pcelevels](https://github.com/mbaudin47/openturns/tree/pcelevels) but I have the following message: 
"Extension error:
L'extension sphinxcontrib.jquery ne peut pas être importée (exception: No module named 'sphinxcontrib.jquery')"

Some remarks on your text: 

- the iso-probabilistic transformation does not necessarily moves to the multivariate normal distribution!
- The polynomial chaos expansion is used with a polynomial basis: we can't talk about the case 'dependent' variables in that doc. I transferred your text on that subject to the more general file 'Functional chaos expansion'.
- I introduced the needs to get a distribution uniquely defined by ll its moements
- you distinguished the easure mu_Z and its pdf.... I kept the word measure mu_Z

We will discuss on it tomorrow!

mbaudin47 commented:
> I would like to commit in [mbaudin47:pcelevels](https://github.com/mbaudin47/openturns/tree/pcelevels) but I have the following message: "Extension error: L'extension sphinxcontrib.jquery ne peut pas être importée (exception: No module named 'sphinxcontrib.jquery')"
I had the same issue. The following command fixed the bug:
```bash
$ pip3 install sphinxcontrib-jquery
```

mbaudin47 commented:
> Thanks for this work! The documentation is really better than the previous one. I would like to commit in [mbaudin47:pcelevels](https://github.com/mbaudin47/openturns/tree/pcelevels) but I have the following message: "Extension error: L'extension sphinxcontrib.jquery ne peut pas être importée (exception: No module named 'sphinxcontrib.jquery')"
> 
> Some remarks on your text:
> 
>     * the iso-probabilistic transformation does not necessarily moves to the multivariate normal distribution!
> 
>     * The polynomial chaos expansion is used with a polynomial basis: we can't talk about the case 'dependent' variables in that doc. I transferred your text on that subject to the more general file 'Functional chaos expansion'.
> 
>     * I introduced the needs to get a distribution uniquely defined by all its moements
> 
>     * you distinguished the measure mu_Z and its pdf.... I kept the word and the notation measure mu_Z
> 
> 
> We will discuss it tomorrow!

mbaudin47 commented:
This is a table presenting the mapping from a distribution to its family of orthogonal polynomials.

![image](https://github.com/openturns/openturns/assets/31351465/2a25cee9-8895-4a76-a955-8d57e1ca5c2f)

josephmure reviewed (commented):
I have answered your replies to my earlier comments

mbaudin47 commented:
@adutfoy : I integrated your changes from https://github.com/mbaudin47/openturns/pull/14. There were several bugs and inconsistencies in the text. Moreover, there is a duplicate content with `FunctionalChaosAlgorithm_doc.i.in`. I fix and merge the content and will submit an updated content afterwards.

mbaudin47 commented:
@josephmure , @adutfoy : This PR seems to be ready from my point of view. Please review it.

mbaudin47 commented on python/src/FunctionalChaosAlgorithm_doc.i.in:
```diff
@@ -203,7 +139,57 @@ Test it:
+...     ot.CorrectedLeaveOneOut()
+... )
+>>> projectionStrategy = ot.LeastSquaresStrategy(inputSample, outputSample, selection)
+>>> algo = ot.FunctionalChaosAlgorithm(inputSample, outputSample, distribution, \
+...                                    adaptiveStrategy, projectionStrategy)"
```
```suggestion
>>> algo = ot.FunctionalChaosAlgorithm( 
...     inputSample, outputSample, distribution, adaptiveStrategy, projectionStrategy
... )"
```

mbaudin47 commented on python/src/FunctionalChaosAlgorithm_doc.i.in:
```diff
@@ -203,7 +139,59 @@ Test it:
+>>> polynomialsList = []
+>>> for i in range(inputDimension):
+...     marginalDistribution = distribution.getMarginal(i)
+...     marginalPolynomial = ot.StandardDistributionPolynomialFactory( \
+...         marginalDistribution
+...     )
```
```suggestion
...     marginalPolynomial = ot.StandardDistributionPolynomialFactory(marginalDistribution)
```

mbaudin47 commented:
@adutfoy  : Please consider this PR, which is now in a consistent state.

adutfoy commented:
Very nice: The doc is really well written and very useful! 
Somme remarks here.


![PC_1](https://github.com/openturns/openturns/assets/13952415/e8a083b4-2703-449d-a789-c470b4c0cda6)

![PC_2](https://github.com/openturns/openturns/assets/13952415/92f86d5b-6491-41cd-842a-7f32826b1c4b)


%%%%%%%%%%%%%%%%%%%%%%%%%
You need to write $\overline{Span (\Psi_k)}$  because $Span (\Psi_k)$ only means finite linear combination of the elements of the basis. 
In the last line, you wanted to write $(\Psi_k)_{k \in \mathbb{N}}$.
%%%%%%%%%%%%%%%%%%%%%%%%%

![PC_3](https://github.com/openturns/openturns/assets/13952415/cc14ac35-cb75-4274-9144-154198eccf49)

%%%%%%%%%%%%%%%%%%%%%%%%%
The sentence in the green box should be put after the second sentence...
%%%%%%%%%%%%%%%%%%%%%%%%%

If you consider the exact expectations, both approaches (scalar product and least-squares pb) are equivalent . But if you discretize the expectations (for ex with quadrature methods or sampling methods), then the approaches are not equivalent: the estimators of the coefficients obtained by both methods are different. So they have different properties (variance).
%%%%%%%%%%%%%%%%%%%%%%%%%

![PC_5](https://github.com/openturns/openturns/assets/13952415/7650796e-26a6-4908-8d3a-384762e7a50f)



![PC_4](https://github.com/openturns/openturns/assets/13952415/1184cddc-22a9-42a3-82a2-25896dec1609)


%%%%%%%%%%%%%%%%%%%%%%%%%
Be carefull: the process you describe is NOT the one that OpenTURNS implements. As a matter of fact, OpenTUR?NS simulates $\tilde{Y}$ thanks to the relation $\tilde{Y} = \tilde{h}(\vect{Z})$ and it does not use $\vect{X}$ any more! This is clearly written in the file FunctionalChaosRandomVector.
%%%%%%%%%%%%%%%%%%%%%%%%%

![PC_6](https://github.com/openturns/openturns/assets/13952415/aa6649d8-9a47-44db-bdc5-dbebf7ef52bd)


%%%%%%%%%%%%%%%%%%%%%%%%%
The $\sqrt{}$ must be removed!
%%%%%%%%%%%%%%%%%%%%%%%%%

adutfoy commented:
![PC_8](https://github.com/openturns/openturns/assets/13952415/316a9e04-7c2c-4381-9c90-13e0f538045c)

%%%%%%%%%%%%%%%%
The first sentences are strange. You should say: 
We consider the isopobabilitic transformation T: X -> Z such that Z has independent components. Thus its pdf can be written as:...
There is a blank bloc afeter the formula...
%%%%%%%%%%%%%%%%
![PC_9](https://github.com/openturns/openturns/assets/13952415/7a177181-078f-4f01-8624-f9ec9865b073)

adutfoy commented:
![PC_10](https://github.com/openturns/openturns/assets/13952415/9f27530b-3c3a-468f-a2ff-22caaea87b86)

mbaudin47 commented:
@adutfoy : Thank you very much for your detailed review. I took all of your comments into account in e6f6cf866. 

I have have a question on the comment that you wrote on the "Functional Chaos Expansion" help page, section "Compute the coefficients". You wrote:

> If you consider the exact expectations, both approaches (scalar product and least-squares pb) are equivalent . But if you discretize the expectations (for ex with quadrature methods or sampling methods), then the approaches are not equivalent: the estimators of the coefficients obtained by both methods are different. So they have different properties (variance).

I agree. Do you want me to insert a section on this topic within the text?

adutfoy commented:
@mbaudin47 : about my remark on the equivalence of both approaches in theory but not inpractice because we approximate $\mathbb{E}$ by empirical mean: yes, you can add this text in the section! I think it will help...

mbaudin47 commented:
> @mbaudin47 : about my remark on the equivalence of both approaches in theory but not in practice because we approximate E by empirical mean: yes, you can add this text in the section! I think it will help...

I will then add a comment on this topic.

mbaudin47 commented:
> @mbaudin47 : about my remark on the equivalence of both approaches in theory but not in practice because we approximate E by empirical mean: yes, you can add this text in the section! I think it will help...

I fixed that in 34d7d55a890b581ccf12ae080c043121808f07b0. To make the point clearer I introduced the discretized versions of the two methods. Please let me know if that corresponds to your requirements.

mbaudin47 commented:
@jschueller : The coverage test fails:
```
1185 - pyinstallcheck_example_plot_estimate_multivariate_distribution (Timeout)
```
but otherwise, this is fine.

mbaudin47 commented:
@adutfoy : I integrated the requested changes. Do you agree on this version?

mbaudin47 commented:
@adutfoy : I integrated the requested changes. Do you agree on this version?

mbaudin47 commented:
@jschueller : I think that this PR is ready for merging.
